File: 53-Maximum Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        nums.insert(nums.begin(), 0);
        for (int i = 1; i < nums.size(); ++i) {
            nums[i] = nums[i] + nums[i - 1];
        }
        
        int minNum = 0, result = nums[1];
        for (int i = 1; i < nums.size(); ++i) {
            if (nums[i] - minNum > result) {
                result = nums[i] - minNum;
            }
            if (minNum > nums[i]) {
                minNum = nums[i];
            }
        }
        return result;
    }
};ʱ

File: 53-Maximum Subarray.java
class Solution {
    public int maxSubArray(int[] nums) {
        int max = Integer.MIN_VALUE, len = nums.length;
        int[] dp = new int[len + 1];
        for (int i = 1; i <= len; ++i) {
            dp[i] = dp[i - 1] > 0 ? dp[i - 1] + nums[i - 1] : nums[i - 1];
            max = Math.max(dp[i], max);
        }
        return max;
    }
}
